File: src-tauri/src/commands/sync.rs
// Tauri IPC commands for account sync
//
// These commands are exposed to the SvelteKit frontend via `invoke()`.
// They handle: auth state, OAuth callback, sync trigger, sync status, logout.

use crate::db::Database;
use crate::sync::{self, auth, SyncState};
use tauri::State;

/// Get the current authentication state (logged in? user profile?).
#[tauri::command]
pub async fn sync_get_auth_state(db: State<'_, Database>) -> Result<auth::AuthState, String> {
    auth::get_auth_state(&db)
}

/// Handle the OAuth callback — store tokens, fetch profile, trigger initial sync.
/// Called from the deep-link handler after `audion://auth/callback?access_token=...&refresh_token=...`
#[tauri::command]
pub async fn sync_handle_auth_callback(
    access_token: String,
    refresh_token: String,
    app_handle: tauri::AppHandle,
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<auth::AuthState, String> {
    tracing::info!("Handling OAuth callback — storing tokens");

    // 1. Store tokens
    auth::store_auth_tokens(&db, &access_token, &refresh_token)?;

    // 2. Fetch user profile from server
    let server_url = sync_state.server_url.lock().unwrap().clone();
    let auth_state =
        auth::fetch_and_store_profile(&db, &server_url, &access_token).await?;

    // 3. Ensure device ID exists
    auth::get_or_create_device_id(&db)?;

    // 4. Trigger initial full sync in background
    let db_clone = db.inner().clone();
    let sync_state_url = sync_state.server_url.clone();
    let is_syncing = sync_state.is_syncing.clone();
    let handle = app_handle.clone();
    tokio::spawn(async move {
        let temp_sync_state = SyncState {
            is_syncing,
            server_url: sync_state_url,
            app_handle: Some(handle),
            provider_mode: std::sync::Arc::new(std::sync::Mutex::new(crate::sync::provider::ProviderMode::Local)),
            sse_join_handle: std::sync::Arc::new(std::sync::Mutex::new(None)),
            is_connected: std::sync::Arc::new(std::sync::atomic::AtomicBool::new(false)),
            db: db_clone.clone(),
        };
        match sync::perform_full_sync(&db_clone, &temp_sync_state).await {
            Ok(_) => tracing::info!("Initial full sync completed"),
            Err(e) => tracing::error!("Initial full sync failed: {}", e),
        }
    });

    Ok(auth_state)
}

/// Log out — revoke the refresh token on the server, clear local auth data.
#[tauri::command]
pub async fn sync_logout(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<(), String> {
    tracing::info!("Logging out — revoking tokens");

    // Try to revoke refresh token on server (best-effort)
    if let Ok(Some(refresh_token)) = auth::get_refresh_token(&db) {
        let body = serde_json::json!({ "refresh_token": refresh_token }).to_string();
        let server_url = sync_state.server_url.lock().unwrap().clone();

        // Fire-and-forget: don't block logout on network issues
        tokio::spawn(async move {
            let client = reqwest::Client::new();
            let _ = client
                .post(format!("{}/auth/logout", server_url))
                .header("Content-Type", "application/json")
                .body(body)
                .send()
                .await;
        });
    }

    // Clear all local auth data + sync queue (best-effort: don't fail logout
    // if the database is missing/corrupted — the frontend will reset UI state regardless)
    if let Err(e) = auth::clear_auth(&db) {
        tracing::warn!(
            "Failed to clear auth data during logout (database may be missing): {}",
            e
        );
    }

    Ok(())
}

/// Trigger a sync — full sync if initial sync hasn't completed, otherwise delta.
#[tauri::command]
pub async fn sync_trigger(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<sync::SyncStatus, String> {
    // Check if logged in first
    let auth_state = auth::get_auth_state(&db)?;
    if !auth_state.is_logged_in {
        return Err("Not logged in".to_string());
    }

    let mode = *sync_state.provider_mode.lock().unwrap();
    if mode == crate::sync::provider::ProviderMode::Server {
        // Custom servers are live streams and do not use the background sync protocol
        return Ok(sync::SyncStatus {
            is_syncing: false,
            last_sync_at: Some(chrono::Utc::now().to_rfc3339()),
            pending_changes: 0,
            last_error: None,
        });
    }

    // If the initial full sync never completed, retry it instead of doing a delta sync
    let full_sync_done = {
        let conn = db.conn.lock().map_err(|e| e.to_string())?;
        crate::db::queries::get_sync_meta(&conn, "full_sync_done")
            .map_err(|e| e.to_string())?
            .map(|v| v == "true")
            .unwrap_or(false)
    };

    if full_sync_done {
        sync::perform_sync(&db, &sync_state).await
    } else {
        tracing::info!("Full sync not yet completed — running full sync instead of delta");
        sync::perform_full_sync(&db, &sync_state).await
    }
}

/// Get the current sync status (pending changes, last sync time, errors).
#[tauri::command]
pub async fn sync_get_status(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<sync::SyncStatus, String> {
    let mode = *sync_state.provider_mode.lock().unwrap();
    if mode == crate::sync::provider::ProviderMode::Server {
        return Ok(sync::SyncStatus {
            is_syncing: false,
            last_sync_at: Some(chrono::Utc::now().to_rfc3339()),
            pending_changes: 0,
            last_error: None,
        });
    }
    sync::get_sync_status(&db, &sync_state)
}

/// Get the server URL for OAuth login (so frontend knows where to open browser).
#[tauri::command]
pub async fn sync_get_server_url(sync_state: State<'_, SyncState>) -> Result<String, String> {
    Ok(sync_state.server_url.lock().unwrap().clone())
}

/// Enqueue a sync change from the frontend (e.g., when a setting changes).
#[tauri::command]
pub async fn sync_enqueue_change(
    entity_type: String,
    entity_id: String,
    operation: String,
    payload: Option<String>,
    db: State<'_, Database>,
) -> Result<(), String> {
    // Only enqueue if logged in
    let auth_state = auth::get_auth_state(&db)?;
    if !auth_state.is_logged_in {
        return Ok(()); // Silently skip if not logged in
    }

    let conn = db.conn.lock().map_err(|e| e.to_string())?;
    crate::db::queries::enqueue_sync_change(
        &conn,
        &entity_type,
        &entity_id,
        &operation,
        payload.as_deref(),
    )
    .map_err(|e| e.to_string())?;

    Ok(())
}

/// Delete the user's account from the server (GDPR).
#[tauri::command]
pub async fn sync_delete_account(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<(), String> {
    tracing::warn!("User requested account deletion (GDPR)");

    let server_url = sync_state.server_url.lock().unwrap().clone();
    auth::authenticated_request(&db, &server_url, "DELETE", "/sync/account", None)
        .await?;

    // Clear local data
    auth::clear_auth(&db)?;

    Ok(())
}

/// Link a Ko-fi donation email to the current account (Flow B — email mismatch).
/// Calls POST /auth/link-kofi via the auth-aware HTTP client (auto-refreshes token).
/// On success, updates locally stored is_supporter + supporter_until and returns
/// the new AuthState so the frontend can update reactively.
#[tauri::command]
pub async fn sync_link_kofi(
    kofi_email: String,
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<auth::AuthState, String> {
    let body = serde_json::json!({ "kofi_email": kofi_email }).to_string();

    let server_url = sync_state.server_url.lock().unwrap().clone();
    let resp_str = auth::authenticated_request(
        &db,
        &server_url,
        "POST",
        "/auth/link-kofi",
        Some(&body),
    )
    .await?;

    // Parse the response to extract the new access token + supporter status
    #[derive(serde::Deserialize)]
    struct LinkKofiResponse {
        access_token: String,
        is_supporter: bool,
        supporter_until: Option<String>,
    }

    let resp: LinkKofiResponse = serde_json::from_str(&resp_str)
        .map_err(|e| format!("Failed to parse link-kofi response: {}", e))?;

    // Store the new access token (it carries updated is_supporter claim)
    {
        let conn = db.conn.lock().map_err(|e| e.to_string())?;
        crate::db::queries::set_sync_meta(&conn, "access_token", &resp.access_token)
            .map_err(|e| e.to_string())?;
    }

    // Update locally stored supporter status
    {
        let conn = db.conn.lock().map_err(|e| e.to_string())?;
        crate::db::queries::set_sync_meta(
            &conn,
            "is_supporter",
            if resp.is_supporter { "true" } else { "false" },
        )
        .map_err(|e| e.to_string())?;
        if let Some(ref until_str) = resp.supporter_until {
            // supporter_until is ISO 8601 string, convert to unix ms
            if let Ok(dt) = chrono::DateTime::parse_from_rfc3339(until_str) {
                let ms = dt.timestamp_millis();
                crate::db::queries::set_sync_meta(&conn, "supporter_until", &ms.to_string())
                    .map_err(|e| e.to_string())?;
            }
        }
    }

    // Return updated auth state
    auth::get_auth_state(&db)
}

/// Get the current access token for WebSocket authentication.
/// If the token is missing or expired, it will attempt to refresh it.
#[tauri::command]
pub async fn sync_get_access_token(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<Option<String>, String> {
    // Check if we have a refresh token (implies we're logged in)
    let rt = auth::get_refresh_token(&db)?;
    if rt.is_none() {
        return Ok(None);
    }

    // Attempt to refresh the access token only if missing or expired.
    let access_token = auth::get_access_token(&db)?;
    if let Some(token) = &access_token {
        if !auth::is_token_expired(token) {
            return Ok(Some(token.clone()));
        }
    }

    // Attempt to refresh the access token to ensure it's valid for the WS connection.
    // If it's already fresh, the server should return a new one or the same one.
    // This is better than returning a potentially expired token and getting a WS 401.
    let server_url = sync_state.server_url.lock().unwrap().clone();
    match auth::refresh_access_token(&db, &server_url).await {
        Ok(token) => Ok(Some(token)),
        Err(e) => {
            tracing::warn!("Failed to refresh token for WebSocket: {}", e);
            // Fallback to current token if refresh fails (might still work if it's not expired)
            Ok(auth::get_access_token(&db)?)
        }
    }
}

/// Get the device ID for identification.
#[tauri::command]
pub async fn sync_get_device_id(db: State<'_, Database>) -> Result<String, String> {
    auth::get_or_create_device_id(&db)
}

#[tauri::command]
pub async fn server_test_connection(
    url: String,
    username: String,
    password: String,
) -> Result<(), String> {
    let url = url.trim_end_matches('/').to_string();

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .map_err(|e| format!("Failed to build HTTP client: {}", e))?;

    let login_url = format!("{}/api/auth/login", url);
    let payload = serde_json::json!({
        "username": username,
        "password": password,
    });

    let resp = client.post(&login_url)
        .json(&payload)
        .send()
        .await
        .map_err(|e| format!("Failed to connect to server: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!("Login failed ({}): {}", status, body));
    }

    #[derive(serde::Deserialize)]
    struct LoginResponse {
        token: String,
        user: UserResponse,
    }
    #[derive(serde::Deserialize)]
    struct UserResponse {
        id: String,
        username: String,
    }

    let _login_res: LoginResponse = resp.json().await
        .map_err(|e| format!("Failed to parse login response: {}", e))?;

    Ok(())
}

#[tauri::command]
pub async fn server_connect(
    url: String,
    username: String,
    password: String,
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<(), String> {
    let url = url.trim_end_matches('/').to_string();

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .map_err(|e| format!("Failed to build HTTP client: {}", e))?;
    let login_url = format!("{}/api/auth/login", url);
    let payload = serde_json::json!({
        "username": username,
        "password": password,
    });

    let resp = client.post(&login_url)
        .json(&payload)
        .send()
        .await
        .map_err(|e| format!("Failed to connect to server: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!("Login failed ({}): {}", status, body));
    }

    #[derive(serde::Deserialize)]
    struct LoginResponse {
        token: String,
        user: UserResponse,
    }
    #[derive(serde::Deserialize)]
    struct UserResponse {
        id: String,
        username: String,
    }

    let login_res: LoginResponse = resp.json().await
        .map_err(|e| format!("Failed to parse login response: {}", e))?;

    auth::store_auth_tokens(&db, &login_res.token, &login_res.token)?;
    auth::store_user_profile(
        &db,
        &login_res.user.id,
        &login_res.user.username,
        Some(&login_res.user.username),
        None,
        true,
        None,
    )?;

    {
        let conn = db.conn.lock().map_err(|e| e.to_string())?;
        crate::db::queries::set_sync_meta(&conn, "server_url", &url).map_err(|e| e.to_string())?;
    }

    *sync_state.server_url.lock().unwrap() = url;
    *sync_state.provider_mode.lock().unwrap() = crate::sync::provider::ProviderMode::Server;
    sync_state.is_connected.store(true, std::sync::atomic::Ordering::SeqCst);

    sync::start_sse_listener(&sync_state, db.inner().clone());

    Ok(())
}

#[tauri::command]
pub async fn server_disconnect(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<(), String> {
    sync::stop_sse_listener(&sync_state);

    auth::clear_auth(&db)?;

    *sync_state.provider_mode.lock().unwrap() = crate::sync::provider::ProviderMode::Local;
    sync_state.is_connected.store(false, std::sync::atomic::Ordering::SeqCst);

    Ok(())
}

#[derive(serde::Serialize)]
pub struct ServerStatus {
    pub connected: bool,
    pub url: String,
    pub user: Option<String>,
}

#[tauri::command]
pub async fn server_get_status(
    db: State<'_, Database>,
    sync_state: State<'_, SyncState>,
) -> Result<ServerStatus, String> {
    let connected = sync_state.is_connected.load(std::sync::atomic::Ordering::SeqCst);
    let url = sync_state.server_url.lock().unwrap().clone();
    
    let user = {
        let conn = db.conn.lock().map_err(|e| e.to_string())?;
        crate::db::queries::get_sync_meta(&conn, "user_name").unwrap_or(None)
    };

    Ok(ServerStatus {
        connected,
        url,
        user,
    })
}

File: src/lib/stores/sync.ts
// Sync state store — manages account sync state for the UI
//
// Provides reactive stores for:
// - Auth state (logged in? user profile?)
// - Sync status (syncing? pending changes? last sync time?)
// - Login/logout actions
// - Sync trigger

import { writable, derived, get } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import { isTauri } from '$lib/api/tauri';
import { refreshAll } from '$lib/stores/library';
import { loadLikedTracks } from '$lib/stores/liked';
import { isOnline } from '$lib/stores/network';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface AuthState {
    is_logged_in: boolean;
    user_id: string | null;
    email: string | null;
    name: string | null;
    avatar_url: string | null;
    is_supporter: boolean;
    /** Unix timestamp in milliseconds. null = no expiry (active subscription). */
    supporter_until: number | null;
}

export interface SyncStatus {
    is_syncing: boolean;
    last_sync_at: string | null;
    pending_changes: number;
    last_error: string | null;
}

export interface SyncProgress {
    phase: string;
    message: string;
    current: number;
    total: number;
}

// ─── Stores ──────────────────────────────────────────────────────────────────

const defaultAuthState: AuthState = {
    is_logged_in: false,
    user_id: null,
    email: null,
    name: null,
    avatar_url: null,
    is_supporter: false,
    supporter_until: null,
};

const defaultSyncStatus: SyncStatus = {
    is_syncing: false,
    last_sync_at: null,
    pending_changes: 0,
    last_error: null,
};

const defaultSyncProgress: SyncProgress = {
    phase: '',
    message: '',
    current: 0,
    total: 0,
};

export const authState = writable<AuthState>(defaultAuthState);
export const syncStatus = writable<SyncStatus>(defaultSyncStatus);
export const syncProgress = writable<SyncProgress>(defaultSyncProgress);
export const showLoginModal = writable<boolean>(false);

// Derived convenience stores
export const isLoggedIn = derived(authState, ($auth) => $auth.is_logged_in);
export const isSupporter = derived(authState, ($auth) => $auth.is_supporter);
export const isSyncing = derived(syncStatus, ($status) => $status.is_syncing);
export const userName = derived(authState, ($auth) => $auth.name || $auth.email || 'User');
export const userAvatar = derived(authState, ($auth) => $auth.avatar_url);

// ─── Event listeners ─────────────────────────────────────────────────────────

let unlistenAuth: UnlistenFn | null = null;
let unlistenSync: UnlistenFn | null = null;
let unlistenDeepLink: UnlistenFn | null = null;
let unlistenProgress: UnlistenFn | null = null;

/**
 * Initialize sync stores — call on app startup.
 * Fetches current auth state and sync status from the Rust backend.
 * Sets up event listeners for real-time updates + deep link handling.
 */
export async function initSync(): Promise<void> {
    if (!isTauri()) return;

    try {
        // Fetch initial state
        const auth = await invoke<AuthState>('sync_get_auth_state');
        authState.set(auth);

        const status = await invoke<SyncStatus>('sync_get_status');
        syncStatus.set(status);
    } catch (error) {
        console.error('[Sync] Failed to initialize:', error);
        // If we can't read auth state (e.g. database was deleted), reset to logged-out
        authState.set(defaultAuthState);
        syncStatus.set(defaultSyncStatus);
    }

    // Listen for auth state changes (e.g., deep-link callback)
    try {
        unlistenAuth = await listen<AuthState>('sync://auth-state-changed', (event) => {
            console.log('[Sync] Auth state changed:', event.payload);
            authState.set(event.payload);
        });

        unlistenSync = await listen<SyncStatus>('sync://status-changed', (event) => {
            console.log('[Sync] Status changed:', event.payload);
            syncStatus.set(event.payload);
            // When sync finishes, reload all data from the now-updated local DB
            if (!event.payload.is_syncing) {
                syncProgress.set(defaultSyncProgress);
                reloadAfterSync();
            }
        });

        unlistenProgress = await listen<SyncProgress>('sync://progress', (event) => {
            console.log(`[Sync] Progress: [${event.payload.phase}] ${event.payload.message} (${event.payload.current}/${event.payload.total})`);
            syncProgress.set(event.payload);
        });
    } catch (error) {
        console.error('[Sync] Failed to set up event listeners:', error);
    }

    // ─── Deep link handling (frontend fallback) ──────────────────────────
    // On some platforms, the Rust-side deep link handler may not fire.
    // Use the official JS deep-link API as a belt-and-suspenders approach.
    try {
        const { onOpenUrl } = await import('@tauri-apps/plugin-deep-link');

        unlistenDeepLink = await onOpenUrl(async (urls: string[]) => {
            console.log('[Sync] Deep link received (JS):', urls);
            for (const url of urls) {
                if (url.startsWith('audion://auth/callback')) {
                    await handleDeepLinkCallback(url);
                }
            }
        });
    } catch (error) {
        console.error('[Sync] Failed to set up deep link listener:', error);
    }

    // Also check if the app was cold-started via a deep link.
    // IMPORTANT: Only process if NOT already logged in.
    // getCurrent() persists across webview refreshes (F5) and would re-login
    // the user immediately after they log out if we didn't guard here.
    try {
        const { getCurrent } = await import('@tauri-apps/plugin-deep-link');
        const currentUrls = await getCurrent();
        if (currentUrls && currentUrls.length > 0) {
            const currentAuth = await invoke<AuthState>('sync_get_auth_state');
            if (!currentAuth.is_logged_in) {
                console.log('[Sync] App cold-started with deep link:', currentUrls);
                for (const url of currentUrls) {
                    if (url.startsWith('audion://auth/callback')) {
                        await handleDeepLinkCallback(url);
                    }
                }
            } else {
                console.log('[Sync] Skipping getCurrent() deep link — user already logged in');
            }
        }
    } catch (error) {
        console.error('[Sync] Failed to check current deep link:', error);
    }


    // ─── Automatic Sync Trigger ──────────────────────────────────────────
    // Watch for pending changes and online status. Trigger sync after a short delay.
    // Enforces a 12-hour cooldown between auto-syncs to reduce server load.
    // Also pauses sync when the app is in the background for > 5 minutes.
    let syncTimeout: ReturnType<typeof setTimeout> | null = null;
    let lastVisibleAt = Date.now();
    let isAppVisible = true;

    if (typeof document !== 'undefined') {
        document.addEventListener('visibilitychange', () => {
            isAppVisible = document.visibilityState === 'visible';
            if (isAppVisible) {
                lastVisibleAt = Date.now();
            }
        });
    }

    const isBackgroundPaused = () => {
        if (isAppVisible) return false;
        const backgroundDuration = Date.now() - lastVisibleAt;
        return backgroundDuration > 5 * 60 * 1000; // 5 minutes
    };
    
    // Automatic Sync Trigger constants
    const LAST_SYNC_KEY = 'audion_last_auto_sync_at';
    const AUTO_SYNC_COOLDOWN_MS = 12 * 60 * 60 * 1000; // 12 hours

    // Prevent initial sync on app open
    let isInitialCheck = true;

    syncStatus.subscribe(($status) => {
        // Skip the very first check to prevent sync on app open
        if (isInitialCheck) {
            isInitialCheck = false;
            console.log('[Sync] Initial check skipped (cooldown active)');
            return;
        }

        const now = Date.now();
        // Read current cooldown timestamp from storage (don't use stale closure variable)
        const currentLastSync = parseInt(localStorage.getItem(LAST_SYNC_KEY) || '0', 10);
        const cooldownRemaining = currentLastSync + AUTO_SYNC_COOLDOWN_MS - now;
        
        const canSync =
            $status.pending_changes > 0 &&
            !$status.is_syncing &&
            get(isOnline) &&
            get(isLoggedIn) &&
            cooldownRemaining <= 0 &&
            !isBackgroundPaused();

        if (canSync) {
            if (syncTimeout) clearTimeout(syncTimeout);
            syncTimeout = setTimeout(() => {
                triggerSync(false);
            }, 5000); // 5 second debounce for auto-sync
        } else if (syncTimeout && ($status.is_syncing || $status.pending_changes === 0)) {
            clearTimeout(syncTimeout);
            syncTimeout = null;
        }
    });

    // Also trigger when coming back online, but still respect cooldown
    isOnline.subscribe(($online) => {
        if ($online && !isInitialCheck) {
            const $status = get(syncStatus);
            const now = Date.now();
            const currentLastSync = parseInt(localStorage.getItem(LAST_SYNC_KEY) || '0', 10);
            if ($status.pending_changes > 0 && 
                get(isLoggedIn) && 
                (now - currentLastSync >= AUTO_SYNC_COOLDOWN_MS) &&
                !isBackgroundPaused()) {
                triggerSync(false);
            }
        }
    });
}

/**
 * Handle a deep link callback URL from the frontend side.
 * Extracts tokens and calls the backend to process them.
 */
async function handleDeepLinkCallback(url: string): Promise<void> {
    try {
        const parsedUrl = new URL(url);
        const accessToken = parsedUrl.searchParams.get('access_token');
        const refreshToken = parsedUrl.searchParams.get('refresh_token');

        if (!accessToken || !refreshToken) {
            console.error('[Sync] Deep link missing tokens');
            return;
        }

        console.log('[Sync] Processing OAuth callback from deep link...');

        const auth = await invoke<AuthState>('sync_handle_auth_callback', {
            accessToken,
            refreshToken,
        });

        authState.set(auth);
        showLoginModal.set(false);

        // The backend spawns a full sync in the background.
        // Poll for completion so the UI updates when it finishes.
        pollSyncStatus();
    } catch (error) {
        console.error('[Sync] Failed to handle deep link callback:', error);
        // Even if the profile fetch failed, the tokens may have been stored.
        // Refresh auth state so the UI reflects whatever state Rust ended up in
        // and close the modal so the user isn't stuck.
        try {
            const auth = await invoke<AuthState>('sync_get_auth_state');
            authState.set(auth);
        } catch (_) {
            // ignore
        }
        showLoginModal.set(false);
    }
}

/**
 * Poll sync status until syncing is complete.
 * Updates the store on each poll so the UI reflects progress.
 */
function pollSyncStatus(intervalMs = 2000, maxAttempts = 60): void {
    let attempts = 0;
    const poll = setInterval(async () => {
        attempts++;
        try {
            const status = await invoke<SyncStatus>('sync_get_status');
            syncStatus.set(status);

            if (!status.is_syncing || attempts >= maxAttempts) {
                clearInterval(poll);
                if (attempts >= maxAttempts) {
                    console.warn('[Sync] Polling timed out after', maxAttempts, 'attempts');
                } else {
                    console.log('[Sync] Sync completed:', status);
                    reloadAfterSync();
                }
            }
        } catch (e) {
            console.error('[Sync] Failed to poll status:', e);
            clearInterval(poll);
        }
    }, intervalMs);
}

/**
 * Reload library, playlists, and liked tracks from the local DB after sync.
 * Called when a background sync finishes so the UI reflects imported data.
 */
async function reloadAfterSync(): Promise<void> {
    try {
        console.log('[Sync] Reloading data after sync...');
        await Promise.all([refreshAll(), loadLikedTracks()]);
        console.log('[Sync] Data reload complete');
    } catch (error) {
        console.error('[Sync] Failed to reload data after sync:', error);
    }
}

/**
 * Clean up event listeners — call on app unmount.
 */
export function destroySync(): void {
    if (unlistenAuth) {
        unlistenAuth();
        unlistenAuth = null;
    }
    if (unlistenSync) {
        unlistenSync();
        unlistenSync = null;
    }
    if (unlistenDeepLink) {
        unlistenDeepLink();
        unlistenDeepLink = null;
    }
    if (unlistenProgress) {
        unlistenProgress();
        unlistenProgress = null;
    }
}

// ─── Actions ─────────────────────────────────────────────────────────────────

/**
 * Start the OAuth login flow — opens the system browser.
 */
export async function startLogin(provider: 'google' | 'github' = 'google'): Promise<void> {
    if (!isTauri()) return;

    try {
        const serverUrl = await invoke<string>('sync_get_server_url');
        const loginUrl = `${serverUrl}/auth/${provider}`;

        // Open in system browser using the opener plugin
        const { openUrl } = await import('@tauri-apps/plugin-opener');
        await openUrl(loginUrl);

        // Close the login modal — the deep-link callback will handle the rest
        showLoginModal.set(false);
    } catch (error) {
        console.error('[Sync] Failed to start login:', error);
    }
}

/**
 * Log out — revoke tokens, clear state.
 */
export async function logout(): Promise<void> {
    if (!isTauri()) return;

    try {
        await invoke('sync_logout');
    } catch (error) {
        console.error('[Sync] Failed to logout:', error);
        // Continue anyway — always clear frontend state regardless of backend errors
        // (e.g. if the local database was deleted, the backend call will fail
        // but we still need to reset the UI to logged-out state)
    }

    // Always reset frontend stores, even if the backend call failed
    authState.set(defaultAuthState);
    syncStatus.set(defaultSyncStatus);
    syncProgress.set(defaultSyncProgress);
}

/**
 * Trigger a sync.
 * @param forced If true, ignores the 12-hour cooldown (manual trigger).
 */
export async function triggerSync(forced = true): Promise<void> {
    if (!isTauri()) return;

    const auth = get(authState);
    if (!auth.is_logged_in) return;

    const LAST_SYNC_KEY = 'audion_last_auto_sync_at';
    const lastSync = parseInt(localStorage.getItem(LAST_SYNC_KEY) || '0', 10);
    const now = Date.now();
    const AUTO_SYNC_COOLDOWN_MS = 12 * 60 * 60 * 1000; // 12 hours

    // Only proceed if forced (manual) or 12 hours passed
    if (!forced && (now - lastSync < AUTO_SYNC_COOLDOWN_MS)) {
        console.log('[Sync] Automatic sync skipped — cooldown active');
        return;
    }

    try {
        console.log(`[Sync] Starting ${forced ? 'manual' : 'scheduled'} sync...`);
        
        // Reset the cooldown timer immediately to prevent loops if this attempt fails
        localStorage.setItem(LAST_SYNC_KEY, now.toString());

        syncStatus.update((s) => ({ ...s, is_syncing: true }));
        syncProgress.set({ phase: 'sync', message: 'Starting sync...', current: 0, total: 0 });
        
        const status = await invoke<SyncStatus>('sync_trigger');
        console.log('[Sync] Sync completed locally:', status);
        
        syncStatus.set(status);
        syncProgress.set(defaultSyncProgress);
        await reloadAfterSync();
    } catch (error) {
        console.error('[Sync] Sync failed:', error);

        let errorMessage = String(error);

        syncStatus.update((s) => ({
            ...s,
            is_syncing: false,
            last_error: errorMessage,
        }));
        syncProgress.set(defaultSyncProgress);
    }
}

/**
 * Enqueue a sync change (called after local mutations).
 */
export async function enqueueChange(
    entityType: string,
    entityId: string,
    operation: string,
    payload?: Record<string, unknown>,
): Promise<void> {
    if (!isTauri()) return;

    try {
        await invoke('sync_enqueue_change', {
            entityType,
            entityId,
            operation,
            payload: payload ? JSON.stringify(payload) : null,
        });

        // Update pending count optimistically
        syncStatus.update((s) => ({
            ...s,
            pending_changes: s.pending_changes + 1,
        }));
    } catch (error) {
        // Don't throw — sync queue failures shouldn't break the app
        console.warn('[Sync] Failed to enqueue change:', error);
    }
}

/**
 * Delete the user's account (GDPR).
 */
export async function deleteAccount(): Promise<void> {
    if (!isTauri()) return;

    try {
        await invoke('sync_delete_account');
    } catch (error) {
        console.error('[Sync] Failed to delete account:', error);
        // Always reset frontend state even on failure
        authState.set(defaultAuthState);
        syncStatus.set(defaultSyncStatus);
        syncProgress.set(defaultSyncProgress);
        throw error;
    }

    // Reset frontend stores on success
    authState.set(defaultAuthState);
    syncStatus.set(defaultSyncStatus);
    syncProgress.set(defaultSyncProgress);
}

/**
 * Refresh auth state from backend (e.g., after returning from background).
 */
export async function refreshAuthState(): Promise<void> {
    if (!isTauri()) return;

    try {
        const auth = await invoke<AuthState>('sync_get_auth_state');
        authState.set(auth);
    } catch (error) {
        console.error('[Sync] Failed to refresh auth state:', error);
        // If we can't read auth state (e.g. database was deleted), reset to logged-out
        authState.set(defaultAuthState);
    }
}

